qiskit/transpiler/passes/scheduling/padding.py

```diff
@@ -0,0 +1,201 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2020.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Padding pass to fill empty timeslot."""
+
+from qiskit.circuit import Qubit
+from qiskit.circuit.delay import Delay
+from qiskit.dagcircuit import DAGCircuit, DAGNode, DAGOutNode
+from qiskit.transpiler.basepasses import TransformationPass
+from qiskit.transpiler.exceptions import TranspilerError
+
+
+class BasePadding(TransformationPass):
+    """The base class of padding pass.
+
+    This pass requires one of scheduling passes to be executed before itself.
+    Since there are multiple scheduling strategies, the selection of scheduling
+    pass is left in the hands of one designs the pass manager.
+    Once schedule pass is called on the DAG circuit, ``time_slot`` is generated
+    in the :attr:`property_set`.  The time slot is the dictionary of
+    the expected instruction execution times keyed on the node instances.
+    Entries in the dictionary are only created for non-delay nodes.
+    The padding pass expects all ``DAGOpNode`` in the circuit to be scheduled in the time slot.
+
+    This pass doesn't define any sequence to interleave, but the pass manages
+    the location where the sequence is inserted, and provides set of information necessary
+    to construct the proper sequence. Thus, a subclass of this pass just need to implement
+    :meth:`_pad` method, in which the subclass constructs a circuit block to insert.
+
+    This mechanism removes lots of boilerplate logic to manage whole DAG circuits.
+    """
+
+    def run(self, dag: DAGCircuit):
+        """Run the measurement alignment pass on `dag`.
+
+        Args:
+            dag (DAGCircuit): DAG to be checked.
+
+        Returns:
+            DAGCircuit: DAG with consistent timing and op nodes annotated with duration.
+
+        Raises:
+            TranspilerError: If the whole circuit or instruction is not scheduled.
+        """
+        if "time_slot" not in self.property_set:
+            raise TranspilerError(
+                f"The input circuit {dag.name} is not scheduled. Call one of scheduling passes "
+                f"before running the {self.__class__.__name__} pass."
+            )
+        time_slot = self.property_set["time_slot"]
+
+        new_dag = DAGCircuit()
+
+        for qreg in dag.qregs.values():
+            new_dag.add_qreg(qreg)
+        for creg in dag.cregs.values():
+            new_dag.add_creg(creg)
+
+        new_dag.name = dag.name
+        new_dag.metadata = dag.metadata
+        new_dag.duration = dag.duration
```

Comment:
I don't think we can remove for backward compatibility because duration is not passed to the new dag and some post program may use duration. Perhaps I misunderstood the mechanism.

Reply:
I was confused because `new_dag.duration` is overwritten in the next line, so this line seems unnecessary.
As for the `duration` property of DAGCircuit (and QuantumCircuit), it is used as a flag if the circuit is scheduled (non-None value) or not (None). So it is sufficient to be set at the last scheduling pass (I'm understanding it should be PadDelay pass).